{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1405.40\n",
    "1. [https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000001405.40/](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000001405.40/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chromosome 22\n",
    "## NC_000022.11 seems to be the same as CM000684.2\n",
    "\n",
    "### Algorithm\n",
    "\n",
    "1. read full sequence at C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\38.p14\\CHROMOSOME-22\n",
    "1. file: nuccore-nc_000022.11.fasta \n",
    "1. size: 50337 Kb \n",
    "1. process file: delete newline chars\n",
    "1. read feature file file:///C:/Users/creeperpandatrex/Documents/1000genomes/DATA/38.p14/awk_ID_symbol_start_end_chr_22\n",
    "1. loop\n",
    "    1. for each line in awk_ID_symbol_start_end_chr_22 \n",
    "    1. read ID, symbol, start, end\n",
    "    1. build the sequence\n",
    "    1. generate hashes: from geneID, from geneSymbol, from gene sequence\n",
    "    1. save as { 'geneID': 'a_number', 'geneSymbol': 'geneSymbolHashXXXXXX', 'geneIDhash':'XXXXXXXX' }\n",
    "    1. and { 'geneIDhash': 'geneHash'}\n",
    "1. Save to a file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GLOBALS\n",
    "import sys\n",
    "import json\n",
    "import gzip\n",
    "from hashlib import sha256\n",
    "import os\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "inputPath = 'C:\\\\Users\\\\creeperpandatrex\\\\Documents\\\\1000genomes\\\\DATA\\\\38.p14\\\\CHROMOSOME-22\\\\'\n",
    "inputFile = 'nuccore-nc_000022.11.fasta'\n",
    "# C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\38.p14\\CHROMOSOME-22\n",
    "# print(os.getenv('SEED'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Delete newlines\n",
    "def removeNewLinez(myPath):\n",
    "    print(myPath)\n",
    "    obj = open(myPath, 'r')\n",
    "    counter = 0\n",
    "    build = ''\n",
    "    print(obj.read(0))\n",
    "    obj.read(10510080)\n",
    "    obj.read(70000)\n",
    "    obj.read(70000)\n",
    "    obj.read(7000)\n",
    "    obj.read(700)\n",
    "    obj.read(1400)\n",
    "    obj.read(140)\n",
    "    obj.read(140)\n",
    "    obj.read(140)\n",
    "    obj.read(140)\n",
    "    obj.read(140)\n",
    "    obj.read(140)\n",
    "    obj.read(140)\n",
    "\n",
    "    \n",
    "    print(obj.read(70))\n",
    "    print(obj.read(70))\n",
    "    print(obj.read(70))\n",
    "\n",
    "    # for x in obj:\n",
    "    #     build += x.strip()\n",
    "    return build\n",
    "myObj = removeNewLinez(inputPath + inputFile)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(myObj))\n",
    "\n",
    "### read from 0 to start\n",
    "\n",
    "print(myObj[18150168])\n",
    "print(myObj[18150169])\n",
    "print(myObj[18150170])\n",
    "print(myObj[18150171])\n",
    "print(myObj[18150172])\n",
    "\n",
    "print()\n",
    "\n",
    "print(myObj[18177390])\n",
    "print(myObj[18177391])\n",
    "print(myObj[18177392])\n",
    "print(myObj[18177393])\n",
    "print(myObj[18177394])\n",
    "print(myObj[18177395])\n",
    "print(myObj[18177396])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Delete newlines\n",
    "def removeNewLinez(myPath):\n",
    "    print(myPath)\n",
    "    obj = open(myPath, 'r')\n",
    "    counter = 0\n",
    "    build = ''\n",
    "    (obj.read(10660000))\n",
    "    (obj.read(12))\n",
    "   \n",
    "    \n",
    "    print(obj.read(70))\n",
    "    print(obj.read(70))\n",
    "    print(obj.read(70))\n",
    "\n",
    "    # for x in obj:\n",
    "    #     build += x.strip()\n",
    "    return build\n",
    "myObj = removeNewLinez(inputPath + inputFile)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Coordinates\n",
    "1. NC_000022.11 at line 150144 : NNNNNNNNNNNNNgaattcttgt\n",
    "1. line 150145: gtttatataa\n",
    "1. USP18 gene: line 259290 in original fasta file : ccccggctcAGT CCC GAC GTG....\n",
    "1. which gives 259289 * 70 = 18150230"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Delete newlines\n",
    "def removeNewLinez(myPath):\n",
    "    print(myPath)\n",
    "    obj = open(myPath, 'r')\n",
    "    counter = 0\n",
    "    build = ''\n",
    "    (obj.read(17894672)) # (25538 * 70 ) + 12\n",
    "    (obj.read(252000))\n",
    "    (obj.read(3200))\n",
    "   \n",
    "    \n",
    "    print(obj.read(210))\n",
    "    print(obj.read(70))\n",
    "    print(obj.read(70)) # difference of 255620 characters, possibly the number of chars in first line\n",
    "    print(obj.read(70))  # it corresponds to 255638 line in the file\n",
    "    print(obj.read(70))\n",
    "\n",
    "    # for x in obj:\n",
    "    #     build += x.strip()\n",
    "    return build\n",
    "myObj = removeNewLinez(inputPath + inputFile)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Delete newlines\n",
    "def removeNewLinez(myPath):\n",
    "    print(myPath)\n",
    "    obj = open(myPath, 'r')\n",
    "    counter = 0\n",
    "    build = ''\n",
    "    print(obj.read(120))\n",
    "\n",
    "    # for x in obj:\n",
    "    #     build += x.strip()\n",
    "    return build\n",
    "myObj = removeNewLinez(inputPath + inputFile)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Something wrong with the reference coordinates. \n",
    "Check the hash\n",
    "\n",
    "Also, apply not just the start/stop coordinates but also match the beginning of the sequence\n",
    "\n",
    "Pending to do the same but with a format where there are not newlines\n",
    "\n",
    "## 50,818,418 characters in chromosome-22 fasta file.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Delete newlines\n",
    "def removeNewLinez(myPath):\n",
    "    print(myPath)\n",
    "    obj = open(myPath, 'r')\n",
    "    counter = 0\n",
    "    build = ''\n",
    "    print(obj.read(120))\n",
    "\n",
    "    for x in obj:\n",
    "        build += x.strip()\n",
    "    return build \n",
    "myObj = removeNewLinez(inputPath + inputFile) ### run on \n",
    "print(len(myObj))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(myObj))\n",
    "idSymbol = [11274, 'USP18']\t\n",
    "start = 18150170\t# real 18150119 in NC_0000022.11\n",
    "end = 18177397\n",
    "# gene FRG1FP\n",
    "idSymbol = [102723780, 'FRG1FP', 119]\n",
    "start = 10940597\n",
    "end = 10961529\n",
    "idSymbol = [503637, 'DUXAP8', -51]\n",
    "start = 15784954\n",
    "end = 15827434\n",
    "size = end - start + 1\n",
    "my = ''\n",
    "counter = 18150000  + 70 + 49 \n",
    "counter = start - 51\n",
    "print(counter)\n",
    "for i in range(size):\n",
    "    my += myObj[counter]\n",
    "    counter += 1\n",
    "\n",
    "print(my)\n",
    "print(type(my))\n",
    "\n",
    "# 50818418\n",
    "# CGCTGTAAGTTTCGCTTTCCATTCAGTGGAAAACGAAAGCTGGGCGGGGTGCCACGAGCGCGGGGCCAGACCAAGGCGGGCCCGGAGCGGAACTTCGGTCCCAGCTCGGTCCCCGGCTCAGTCCCGACGTGGAACTCAGCAGCGGAGGCTGGACGCTTGCATGGCGCTTGAGGCAAGTTCGGGGCTCATTTTGGAAGTTTTCTTTCTAGC\n",
    "# considerar add one or two nucleotides to the ends, maybe 2 or 3 to have 2 or 3 hashes with only one or\n",
    "# two nucleotide differences\n",
    "# pending to decide how to deal with the ncbi reported gene and the actual sequence derived from the \n",
    "# reference genome with reported, although displaced coordinates\n",
    "# Build the gene sequence, use 5', as well as 3' end primers to confirm taken from the reported by ncbi\n",
    "# not by just finding them in the reference-genome.fasta file."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "fox-conda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
