# Procedure overview

## 9.8.2024

1. 38.p14 full assembly n = 100
1. individual compressed chromosomes n = 80
1. default n in samtools consensus = 70
1. samtools some options:
```
(samtools-conda-wsl-create) creeper@creepypandatrex:~/test01$ cat command
samtools consensus -a --show-ins no --line-len 100 /mnt/c/Users/creeperpandatrex/Documents/1000genomes/HG0040X.final.cram -o HG0040X.v3.fa
samtools consensus -a --show-ins no --line-len 100 -r chr1 /mnt/c/Users/creeperpandatrex/Documents/1000genomes/HG00408.final.cram -o HG00408.chr1.v3.fa
samtools consensus -a --show-ins no --line-len 100 -r chr22:41469117-41528974 HG00405.final.cram -o HG00405.chr22.50-ACO2.fa
(samtools-conda-wsl-create) creeper@creepypandatrex:~/test01$
```
1. tested -r region; --ambig; --line-len; whole chromosome, or part of a chromosome
1. Regarding coordinates: divide the start by N, add 1 because of header, the fraction by 100 is the percent of bases for the first nucleotide in the start position. 
1. In the reference genome it worked. In the extracted 405 the diff is different for 2 loci out of 3 by 1500 and 7000 nucleotide, hence, unreliable. 
1. Extracted genes '5902'-RANBP1 and '50'-ACO2 from 403, 404, 405 cram files.
1. First with N, then with option --ambig
1. Quite many R, W, S, K found.
1. Pending the count on the selected loci, then think about what to do going forward: Option a: using samtools to extract sequences based on the ref sequence. To many genes/loci to run the command. OPTION B: run consensus, check each loci by coordinates, doble check by primers, even though there are quite many NNNNNNN runs, as well as Ns in the target sequences, maybe confirm the char position to run the hash on the sequence with IUPAC codes, and/or, hash the number of consensus depending on the IUPAC code, remember that at each loci there at least two variant on each nucleotide position. If het/het expect the segregation, if homo/het, expect the segregation, if homo/homo, then only one possible variation in the child. 
## 9.7.2024 (m.d.y)

1. 405 child of 405(father) and 403(mother)
1. Father ERR3241665 and mother ERR3241666
1. ERR3988761 - HG00405.final.cram
1. ERR3988762 - HG00408.final.cram
1. ERR3988763 - 418
1. Python env: 
```(samtools-conda-wsl-create) creeper@creepypandatrex:~/test01$ samtools consensus -a --show-ins no /mnt/c/Users/creeperpandatrex/Documents/1000genomes/HG00405.final.cram -o HG00405.v3.fa
(samtools-conda-wsl-create) creeper@creepypandatrex:~/test01$
```
1. Pending to test --show-ins yes
1. wc HG00405.v3.fa:
```
(base) creeper@creepypandatrex:~/test01$ wc HG00405.v3.fa
  45934179   45934179 3261099308 HG00405.v3.fa
(base) creeper@creepypandatrex:~/test01$
```
1. grep '>' HG00405.v3.fa | wc -l  = 2805
1. 3,261,099,308 chars
1. samtools view -H
```sh
samtools view -H in.cram
```
1. Preliminar result of generating the consensus good
1. However, coordinate mismatch of about 50,000 nucleotides
1. gene 57185, NIPAL3, start: 24,413,524; end: 24,472,983; length 59460 nucl

## 9.6.2024 (m.d.y)

### Samtools install
1. tried in windows python env, it did not work
1. Switched to wsl ubuntu
1. Install miniconda3, search was suggesting mamba
1. install miniconda3, export the path, created samtools-conda-wsl-create environment
1. conda config --add channels bioconda and conda-forge
1. conda activate samtools-conda....
1. conda install -c bioconda samtools
1. run 
```
(samtools-conda-wsl-create) creeper@creepypandatrex:~/test01$ samtools fasta /mnt/c/Users/creeperpandatrex/Documents/1000genomes/HG00405.final.cram > HG00405.fa
```
1. After two hours it generated a 123 GBytes file !!!!!
1. It created a .cache from web or internal md5 tags
1. The reference assembly is 3 GB only
1. Fasta seems to be extracting short reads only.
1. I downloaded samtools/misc/seq_cache_populate.pl script and used Homo_sapiens_assembly.38.fasta about 3 GB
1. It generated the tree as /cache-root/%2s/%2s/%s to be written into REF_CACHE
1. diff /f9/8d/b........... on the downloaded version and the generated by script: identical
1. renamed the downloaded version after stopping samtools and moved the generated full tree
1. Restarted the samtools fasta command and generated about 8GB in just few minutes.
1. But it looks that it is not the consensus a the sequences from NGS
1. Try consensus or view or mpileup to generate a full assembly.
1. During the night restart the downloading of the cache to make sure it correspond since the error after 8GB download was:
```
(samtools-conda-wsl-create) creeper@creepypandatrex:~/test01$ samtools fasta /mnt/c/Users/creeperpandatrex/Documents/1000genomes/HG00405.final.cram > HG00405.fa
+-[E::bgzf_read_block] Failed to read uncompressed data at offset 190850128: Input/output error
[E::bgzf_read] Read block operation failed with error 4 after 6730 of 44772 bytes
bgzf_read() on reference file: Input/output error
[E::cram_decode_slice] Unable to fetch reference #0:190843399-190888170

[E::cram_next_slice] Failure to decode slice
samtools bam2fq: Failed to read bam record
samtools bam2fq: Error writing to FASTx files.: Read-only file system
[M::bam2fq_mainloop] discarded 0 singletons
[M::bam2fq_mainloop] processed 46732642 reads
(samtools-conda-wsl-create) creeper@creepypandatrex:~/test01$ samtools fasta /mnt/c/Users/creeperpandatrex/Documents/1000genomes/HG00405.final.cram > HG00405.fa
```
1. The slices were different, maybe different number of Ns or something.



### from biostars 367,960
```
The answer to this question is a simple two-step workflow:

Call variants
Include the variants into the reference sequence
Do the variant calling step with your favorite program/workflow, e.g. with bcftools:

    $ bcftools mpileup -Ou -f ref.fa input.bam | bcftools call -Ou -mv | bcftools norm -f ref.fa -Oz -o output.vcf.gz
In the end, one needs a valid, sorted vcf file which is compressed with bgzip and indexed with tabix.

I also recommend to normalize your vcf file (In the above command this is done by bcftools norm).

If your vcf isn't already compressed you can do this with:

    $ bgzip -c output.vcf > output.vcf.gz
And indexing is done by:

    $ tabix output.vcf.gz
Now we are ready to create our consensus sequence. The tool of choice is bcftools consensus.

If one like to create a complete new reference genome, based on the called variants, this is done by:

    $ bcftools consensus -f ref.fa output.vcf.gz > out.fa
If you are interested in only a given region:

    $ samtools faidx ref.fa 8:11870-11890 | bcftools consensus output.vcf.gz -o out.fa
You can also create consensus sequences for multiple regions, if you provide a bed file to samtools faidx:

    $ samtools faidx -r regions.bed ref.fa  | bcftools consensus output.vcf.gz -o out.fa
Have fun 
```
1. https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/  2024
1. and
1. https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/  2015 ????

## 9.5.2024 

1. How to get individual chromosome sequence in fasta format from vcf.gz and its vcf.gz.tbi file of 1000 genome project?
1. https://www.biostars.org/p/172109/
1. How to make consensus sequence from aligned reads ? https://www.biostars.org/p/9465132/
1. https://www.biostars.org/p/367960/ : explains the steps to get consensus using bcftools
1. create consensus sequence from bam file https://www.biostars.org/p/419927/
1. VCF - Variant Call Format see https://davetang.github.io/learning_vcf_file/

## 9.2.2024 (m.d.y)

1. Telomere-to-telomere project at https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_009914755.1/
1. Pending to consider whether it is important going forward. 
1. The coordinate system looks different.
1. Count stream of chars from a file: using vi, colon 'goto position number + position/80 + numof chars in '>dsfdffirst line"
1. Is the start of the gene always the first exon? apparently yes
1. Advantages of hashes faa vs fna from dDNA?. Pending
1. Rerun the program with the correct gene coordinates (+- 11-12 chars: numofchar in first line)
1. Sanity check
```sh
$ for i in *.primer; do echo '============ HEAD === ' $i '============'; zless $i | head; echo '------------------ HEAD ----'; echo '--------------- TAIL ---------'; zless $i | tail; echo '--------------- TAIL '; done > heads-tails
```
1. TEST-chr10: 
```
>('338591', 'CHCHD3P1', '8619806', '8620809')	1004
GCTCCTGCTCTGGGGGAAAAGAATCCAGGCCCCTCCACGC....ATTGGCAGCAATTTGAAGGGCGGCCAAAACCTAGATCAGG
```
1. insilico PCR at https://genome.ucsc.edu/cgi-bin/hgPcr?hgsid=2341707234_NoL7BXYigD5mZhzNggr6pXJO7LVR&org=Human&db=hg38&wp_target=genome&wp_f=GCTCCTGCTCTGGGGGAAAAGAATCCAGGCCCCTCCACGC&wp_r=ATTGGCAGCAATTTGAAGGGCGGCCAAAACCTAGATCAGG&Submit=Submit&wp_size=4000&wp_perfect=15&wp_good=15&wp_flipReverse=on&boolshad.wp_flipReverse=0&wp_append=on&boolshad.wp_append=0
1. OUTPUT:  	
    >chr10:8619818+8620821 1004bp GCTCCTGCTCTGGGGGAAAAGAATCCAGGCCCCTCCACGC CCTGATCTAGGTTTTGGCCGCCCTTCAAATTGCTGCCAAT
1. Reference to fix map and patches: 
    1. https://www.ncbi.nlm.nih.gov/grc/help/definitions/#ALTERNATE
    1. https://genome-blog.gi.ucsc.edu/blog/2019/02/22/patches/
    1. https://genome.ucsc.edu/FAQ/FAQdownloads#downloadAlt
1. ref: https://genome.ucsc.edu/cgi-bin/hgc?hgsid=2341717678_FPa0eEaqsWn07YckYuaXgoP5mUpH&g=htcGetDna2&table=hgnc&i=HGNC%3A44697&o=8619805&l=8619805&r=8620808&getDnaPos=chr10%3A8%2C619%2C806-8%2C620%2C808&hgSeq.cdsExon=1&hgSeq.padding5=0&hgSeq.padding3=0&hgSeq.casing=upper&boolshad.hgSeq.maskRepeats=0&hgSeq.repMasking=lower&boolshad.hgSeq.revComp=0&submit=Get+DNA
1. OUTPUT:
```
>hg38_hgnc range=chr10:8619806-8620808 5'pad=0 3'pad=0 strand=+ repeatMasking=none 
1        1         2         3         4         5
GGTCGTGGCCTTGCTCCTGCTCTGGGGGAAAAGAATCCAGGCCCCTCCAC...ATCCAACAGAATCATACCACATTTCCATTGGCAGCAATTTGAAGGGCGGC..CAA
            ||||||||||||||||||||||||||||||||||||||                             ||||||||||||||||||||||||///
            GCTCCTGCTCTGGGGGAAAAGAATCCAGGCCCCTCCACGC....                 ......ATTGGCAGCAATTTGAAGGGCGGCCAAAACCTAGATCAGG
TEST-chr10
```
1. At NCBI : https://www.ncbi.nlm.nih.gov/gene/338591
1. Web page:
```
Location: 10p14
Annotation release	Status	Assembly	Chr	Location
RS_2024_08	current	GRCh38.p14 (GCF_000001405.40)	10	NC_000010.11 (8619806..8620809)
RS_2024_08	current	T2T-CHM13v2.0 (GCF_009914755.1)	10	NC_060934.1 (8621279..8622282)
105.20220307	previous assembly	GRCh37.p13 (GCF_000001405.25)	10	NC_000010.10 (8661769..8662772)
```
1. OUTPUT: https://www.ncbi.nlm.nih.gov/nuccore/NC_000010.11?report=fasta&from=8619806&to=8620809
```
>NC_000010.11:8619806-8620809 Homo sapiens chromosome 10, GRCh38.p14 Primary Assembly
1        1         2         3         4         5         6         7
GGTCGTGGCCTTGCTCCTGCTCTGGGGGAAAAGAATCCAGGCCCCTCCACGCGAGTGTGGGCACGGGGGA...GCAGCAATTTGAAGGGCGGCCAAA
            ||||||||||||||||||||||||||||||||||||||||                     ||||||||||||||||||||||||
            GCTCCTGCTCTGGGGGAAAAGAATCCAGGCCCCTCCACGC.................ATTGGCAGCAATTTGAAGGGCGGCCAAAACCTAGATCAGG
            TEST-chr10
```
1. Test-chr3
```
>('90102', 'PHLDB2', '111732496', '111976517')	244022
GCAGAGGCCAGAGAGAGCAGGAAAGGAAATGGAAAGGAAC....AGATACATTTATTGCTGGGGATTGCAAGGAGTGTGTTTAC
```
1. at NCBI
```
Location: 3q13.2
Exon count: 21
Annotation release	Status	Assembly	Chr	Location
RS_2024_08	current	GRCh38.p14 (GCF_000001405.40)	3	NC_000003.12 (111732496..111976517)
RS_2024_08	current	T2T-CHM13v2.0 (GCF_009914755.1)	3	NC_060927.1 (114453343..114697610)
105.20220307	previous assembly	GRCh37.p13 (GCF_000001405.25)	3	NC_000003.11 (111451343..111695364)
```
1. NCBI
```
>NC_000003.12:111732496-111976517 Homo sapiens chromosome 3, GRCh38.p14 Primary Assembly
1        1         2         3         4         5         6         7   1        1         2         3         4         5         6         7
GTGACCACAGTCTGCAGAGGCCAGAGAGAGCAGGAAAGGAAATGGAAAGGAACCTCACCTTCATGCTTGG...ATGAACCCTGAAAAACTCTGACATCATATTGTTGTCTGTTTACACAGATACATTTATTGCTGGGGATTGC..AA
             ||||||||||||||||||||||||||||||||||||||||                                                                 |||||||||||||||||||||||||//
             GCAGAGGCCAGAGAGAGCAGGAAAGGAAATGGAAAGGAAC.................................................................AGATACATTTATTGCTGGGGATTGCAAGGAGTGTGTTTAC
```
1. Test-chr22
```
>('3162', 'HMOX1', '35381096', '35394207')	13112
CTCTCGAGCGTCCTCAGCGCAGCCGCCGCCCGCGGAGCCA....AAATAATAAACAACATTGTCTGATAGTAGCTTGAAGTAGT
>NC_000022.11:35381096-35394207 Homo sapiens chromosome 22, GRCh38.p14 Primary Assembly
1        1         2         3         4         5         6         7   4         5         6         7
AACGCCTGCCTCCTCTCGAGCGTCCTCAGCGCAGCCGCCGCCCGCGGAGCCAGCACGAACGAGCCCAGCA...GTGGCCTTGGTCTAACTTTTGTGTGAAATAA..TAAACAACATTGTCTGATAGTA
            ||||||||||||||||||||||||||||||||||||||||                                                      ||||||||||||||||||||||           
            CTCTCGAGCGTCCTCAGCGCAGCCGCCGCCCGCGGAGCCA................................................AAATAATAAACAACATTGTCTGATAGTAGCTTGAAGTAGT


```
1. I found the reason of the mismatch: number of chars first line + number of start/80 (meaning newline chars). Vi allows to jump to start or end but it needs added the number of char in first line plus number of newline chars.
```
1        1         2         3         4         5         6         7
>NC_000010.11 Homo sapiens chromosome 10, GRCh38.p14 Primary Assembly 71
>NC_000011.10 Homo sapiens chromosome 11, GRCh38.p14 Primary Assembly 71
>NC_000012.12 Homo sapiens chromosome 12, GRCh38.p14 Primary Assembly 71
>NC_000013.11 Homo sapiens chromosome 13, GRCh38.p14 Primary Assembly 71
>NC_000015.10 Homo sapiens chromosome 15, GRCh38.p14 Primary Assembly 
>NC_000016.10 Homo sapiens chromosome 16, GRCh38.p14 Primary Assembly
>NC_000017.11 Homo sapiens chromosome 17, GRCh38.p14 Primary Assembly
>NC_000018.10 Homo sapiens chromosome 18, GRCh38.p14 Primary Assembly
>NC_000019.10 Homo sapiens chromosome 19, GRCh38.p14 Primary Assembly
>NC_000022.11 Homo sapiens chromosome 22, GRCh38.p14 Primary Assembly 71
>NC_000020.11 Homo sapiens chromosome 20, GRCh38.p14 Primary Assembly 71
>NC_000014.9 Homo sapiens chromosome 14, GRCh38.p14 Primary Assembly 70
>NC_000001.11 Homo sapiens chromosome 1, GRCh38.p14 Primary Assembly 70
>NC_000002.12 Homo sapiens chromosome 2, GRCh38.p14 Primary Assembly
>NC_000021.9 Homo sapiens chromosome 21, GRCh38.p14 Primary Assembly
>NC_000003.12 Homo sapiens chromosome 3, GRCh38.p14 Primary Assembly
>NC_000004.12 Homo sapiens chromosome 4, GRCh38.p14 Primary Assembly
>NC_000005.10 Homo sapiens chromosome 5, GRCh38.p14 Primary Assembly
>NC_000006.12 Homo sapiens chromosome 6, GRCh38.p14 Primary Assembly
>NC_000007.14 Homo sapiens chromosome 7, GRCh38.p14 Primary Assembly
>NC_000008.11 Homo sapiens chromosome 8, GRCh38.p14 Primary Assembly
>NC_000009.12 Homo sapiens chromosome 9, GRCh38.p14 Primary Assembly
>NC_000023.11 Homo sapiens chromosome X, GRCh38.p14 Primary Assembly
>NC_000024.10 Homo sapiens chromosome Y, GRCh38.p14 Primary Assembly
```
1. Find in python notebook, cell with processChrWithIdSymbolSanityCheck in def buildSeqHash(tuple, mySeed): instead of hardcoded 80 pass the real length of the header line + newLine char. in cell 189 of chrNN.fna.gz.ipynb file
1. ### Pending from chrNN.fna.gz.ipynb cell 193
1. 6.23.2024
1. .primer is two lines: first line with > and tuple with id, symbol, start, end
1. .primer second line with gene 5'primer 3'end primer 
1. direction based in reference sequence
1. Checked 
```
 19 >('5902', 'RANBP1', '20116104', '20127355')     11252
     20 ACTGTCCATAGAGGGGTCACCACGTCGGCCACTCGTGTTA....CATCAAAAATAGTGAATAAAAAACACATTTGGAACCTGGG
     >NC_000022.11:20116104-20127355 Homo sapiens chromosome 22, GRCh38.p14 Primary Assembly
ACTGTCCATAGAGGGGTCACCACGTCGGCCACTCGTGTTACTGGTGGCTCACTCTCACCCTCCTGTCGCC...CTTATTTATAGTCATCAAAAATAGTGAATAAAAAACACATTTGGAACCTGGG
||||||||||||||||||||||||||||||||||||||||                                             ||||||||||||||||||||||||||||||||||||||||
ACTGTCCATAGAGGGGTCACCACGTCGGCCACTCGTGTTA.............................................CATCAAAAATAGTGAATAAAAAACACATTTGGAACCTGGG
  
```
And it looks good
1. NCBI at 5902
```
Location: 22q11.21
Exon count: 9
Annotation release	Status	Assembly	Chr	Location
RS_2024_08	current	GRCh38.p14 (GCF_000001405.40)	22	NC_000022.11 (20116104..20127355)
RS_2024_08	current	T2T-CHM13v2.0 (GCF_009914755.1)	22	NC_060946.1 (20495154..20506390)
105.20220307	previous assembly	GRCh37.p13 (GCF_000001405.25)	22	NC_000022.10 (20103627..20114878)
```

## 8.31.2024 (m.d.y)


### Processing faa

1. It was relatively easy: convert FASTA to 2 lines per protein by deleting new line chars from FASTA format
1. file, key/value: hash-label/hash-sequence
1. file, key/value: label/hash-label
1. count hash-sequence duplicates
1. file, key/['values-hash-label',]: hash-sequence/['hash-label1', 'hash-label2','etc'.....']
1. Some sequences highly repeated in GCG_versions, most sequence  unique

### Processing fna genomic data GRCh38.p14

1. https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GCF_000001405.40_GRCh38.p14_assembly_structure/Primary_Assembly/assembled_chromosomes/FASTA/
1. Example : >NC_000022.11 Homo sapiens chromosome 22, GRCh38.p14 Primary Assembly
1. creeperpandatrex@creepypandatrex MINGW64 ~/Documents/1000genomes/DATA/GCF000001405-38p14-assembly
1.  zgrep '>' chr22.fna.gz
    ```>NC_000022.11 Homo sapiens chromosome 22, GRCh38.p14 Primary Assembly
    $ zcat chr22.fna.gz | wc -l
    635232 lines
    Each line 80 characters```
1. 635232 x 80 = 50,818,560 characters
1. file compressed size: 10.4 MB
1. zgrep -P 'chromosome\t22' in 38.p14_feature_table to extract features for chromosome 22
1. Convert FASTA format to 2 lines format.
1. Get the features from feature_table file
```
$ zgrep -P 'chromosome\t22' GCF_000001405.40_GRCh38.p14_feature_table.txt.gz | awk -F'\t' '{print$1}' | sort --unique
CDS
C_region
J_segment
V_segment
gene
mRNA
misc_RNA
ncRNA
precursor_RNA
rRNA
tRNA
```
1. zgrep 'chromosome 22' => 4357 hits
1. zgrep 'chromosome 22' + 'gene' => 896 hits
1. file: [ID, symbol, start, end]
1. read [ID, symbol, start, end] as set to eliminate duplicates
1. with chrNN.fna.gz and [ID, symbol, start, end] as in <img src="./images/Screenshot 2024-06-22 064723.png"/>
1. hash the [id, symbol, start, end].
1. build the sequence between start-end and hash
1. save as [{},{},{key/val}, {idSymbolHash: builtSeqHash}, {}, {},...{}]
1. file labelmap: key/value, hash/[id, symbol, start, end]
1. file labelseqhash: key/value, idSymbolHash/seqHash
1. file save as compressed chrNN.fna.gz.labelmap and labelseqhash

1. at some point do electronic PCR using primers on the 3' and 5' ends of the gene base on the feature_table on individual sequences


