# Genome Assembly Analysis

1. Chromosome 1
    1. fasta2one 19 minutes
    1. Total: 31108 gene
    1. Unique: 28828 genes
1. Chromosome 2
    1. fasta2one 22 minutes
    1. Total: 8745 gene
    1. Unique: 7997 genes
1. Chromosome 3
    1. fasta2one 15 minutes
    1. Total: 3628 gene
    1. Unique: 3393 genes
1. Chromosome 4
    1. fasta2one 15 minutes
    1. Total: 2879 gene
    1. Unique: 2760 genes
1. Chromosome 5
    1. fasta2one 14 minutes
    1. Total: 3362 gene
    1. Unique: 2958 genes
1. Chromosome 6
    1. fasta2one 19 minutes
    1. Total: 3699 gene
    1. Unique: 3451 genes
1. Chromosome 7
    1. 10 min
    1. Total: 3406
    1. Unique: 3111
1. 8
    1. 7 minutes
    1. 2596
    1. 2444
1. 9
    1. 7 min
    1. 2734
    1. 2584
1. 10
    1. 7 min
    1. 2726
    1. 2427
1. 11
    1. 7 min
    1. 3468
    1. 3235
1. 12
    1. 7
    1. 3070
    1. 2853
1. 13
    1. 5
    1. 1591
    1. 1542
1. 14
    1.  5 min
    1. 2372
    1. 2232
1. 15
    1. 4 min
    1. 2257
    1. 2040
1. 16
    1. 3 min
    1. 2408
    1. 2231
1. 17
    1. 3 min
    1. 2987
    1. 2728
1. 18
    1. 3 min
    1. 1156
    1. 1118
1. 19
    1. 2 min
    1. 2921
    1. 2701
1. 20
    1. 2 min
    1. 1661
    1. 1488
1. 21
    1. 1 min
    1. 896
    1. 854
1. 22
    1. 1 min
    1. 1523
    1. 1329
1. Chromosome X 24 
    1. 10 min
    1. 2512
    1. 2375 
1. Chromosome Y 25
    1. 2 min
    1. 611
    1. 590

## Added to karysim/public/data/CHROMOSOME folder

### 6.28.2024

All files chr*.fna.gz.label*.38.p14 files corresponding to each chromosome. 

Chromosome X is 24 number
Chromosome Y is 25 number. 

## PENDING

1. Find how to extract consensus from cram files
1. Loop from 1 through 25: about 3 hours
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00403/
1. Raw data as fastq.gz files pair readings multiple
1. CHECK CHECK CHECK CHECK CHECK CHECK CHECK CHECK !!!!!!! 6.29.2024
1. CHECK CHECK CHECK CHECK CHECK CHECK CHECK CHECK !!!!!!! 6.29.2024
1. CHECK CHECK CHECK CHECK CHECK CHECK CHECK CHECK !!!!!!! 6.29.2024
1. CHECK CHECK CHECK CHECK CHECK CHECK CHECK CHECK !!!!!!! 6.29.2024
1. CHECK CHECK CHECK CHECK CHECK CHECK CHECK CHECK !!!!!!! 6.29.2024
1. CHECK CHECK CHECK CHECK CHECK CHECK CHECK CHECK !!!!!!! 6.29.2024
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/additional_698_related/1000G_698_related_high_coverage.sequence.index
1. ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR398/ERR3988761/HG00405.final.cram	25c1561b4f603c5c0e763d9d7b63f0d2	ERR3988761	ERP120144	30X whole genome sequencing coverage of 698 samples to complete trios in Phase 3 1000 Genome sample set	NYGC	ERA2409803	2020-03-13 00:00	ERS4367795	HG00405	CHS		ERX3990626	ILLUMINA	Illumina NovaSeq 6000	HG00405	ena-RUN-NYGC-13-03-2020-19:45:34:171-1	450	PAIRED		730496712	109574506800	high_cov
1. 
ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR398/ERR3988762/HG00408.final.cram	abb848a98ece4f07f1b4a5977e9728be	ERR3988762	ERP120144	30X whole genome seq-uencing coverage of 698 samples to complete trios in Phase 3 1000 Genome sample set	NYGC	ERA2409803	2020-03-13 00:00	ERS4367796	HG00408	CHS		ERX3990627	ILLUMINA	Illumina NovaSeq 6000	HG00408	ena-RUN-NYGC-13-03-2020-19:45:34:171-2	450	PAIRED		755902162	113385324300	high_cov
f
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/release/20130502/supporting/high_coverage_alignments/20141118_high_coverage.alignment.index
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/release/20130502/supporting/high_coverage_alignments/

    1. ftp/phase3/data/HG00419/high_coverage_alignment/HG00419.wgs.ILLUMINA.bwa.CHS.high_cov_pcr_free.20140203.bam	5e98e713f0ecb42ffbb11a1f21053f69	
    1. ftp/phase3/data/HG00419/high_coverage_alignment/HG00419.wgs.ILLUMINA.bwa.CHS.high_cov_pcr_free.20140203.bam.bai	3c298f84c7e768cbf34535d56b5c3d18
    1. ftp/phase3/data/HG00096/high_coverage_alignment/HG00096.wgs.ILLUMINA.bwa.GBR.high_cov_pcr_free.20140203.bam	7e24df4489ba5ac7d27d188547cf885d	
    1. ftp/phase3/data/HG00096/high_coverage_alignment/HG00096.wgs.ILLUMINA.bwa.GBR.high_cov_pcr_free.20140203.bam.bai	d2e7660be3979ea3495f9a482406f512
    1. ftp/phase3/data/NA20502/high_coverage_alignment/NA20502.wgs.ILLUMINA.bwa.TSI.high_cov_pcr_free.20140203.bam	7ab29405614376b80391da88d3f094eb	ftp/phase3/data/NA20502/high_coverage_alignment/NA20502.wgs.ILLUMINA.bwa.TSI.high_cov_pcr_free.20140203.bam.bai	04daa3fe339326df3e9b06ff403f1ad1
    1. ftp/phase3/data/NA20845/high_coverage_alignment/NA20845.wgs.ILLUMINA.bwa.GIH.high_cov_pcr_free.20140203.bam	b6330dc3a509c2fbbb506f0ee2ce0767	
    1. ftp/phase3/data/NA20845/high_coverage_alignment/NA20845.wgs.ILLUMINA.bwa.GIH.high_cov_pcr_free.20140203.bam.bai	0c32671de79b056d575a703c3ddba2b9
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00403/

    1. alignment/                    2024-06-11 17:30    -   
    1. cg_data/                      2024-06-11 17:30    -   
    1. exome_alignment/              2024-06-11 17:30    -   
    1. sequence_read/                2024-06-11 17:30    -   

1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00403/

1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00403/alignment/
    1. HG00403.chrom11.ILLUMINA.bwa.CHS.low_coverage.20120522.bam          2012-11-15 22:07  551M  
    1. HG00403.chrom11.ILLUMINA.bwa.CHS.low_coverage.20120522.bam.bai      2012-11-15 22:07  373K  
    1. HG00403.chrom11.ILLUMINA.bwa.CHS.low_coverage.20120522.bam.bas      2012-11-26 17:23  1.0K  
    1. HG00403.chrom20.ILLUMINA.bwa.CHS.low_coverage.20120522.bam          2012-11-15 22:07  224M  
    1. HG00403.chrom20.ILLUMINA.bwa.CHS.low_coverage.20120522.bam.bai      2012-11-15 22:07  169K  
    1. HG00403.chrom20.ILLUMINA.bwa.CHS.low_coverage.20120522.bam.bas      2012-11-26 17:23  1.0K  
    1. HG00403.mapped.ILLUMINA.bwa.CHS.low_coverage.20120522.bam           2012-11-15 22:20   12G  
    1. HG00403.mapped.ILLUMINA.bwa.CHS.low_coverage.20120522.bam.bai       2012-11-15 22:20  8.1M  
    1. HG00403.mapped.ILLUMINA.bwa.CHS.low_coverage.20120522.bam.bas       2012-11-26 17:21  1.1K  
    1. HG00403.mapped.ILLUMINA.bwa.CHS.low_coverage.20120522.bam.cram      2013-06-29 12:08  4.5G  
    1. HG00403.mapped.ILLUMINA.bwa.CHS.low_coverage.20120522.bam.cram.crai 2013-06-29 12:12  199K  
    1. HG00403.mapped.ILLUMINA.bwa.CHS.low_coverage.20120522.bam.csra      2013-04-17 23:25  4.3G  
    1. HG00403.unmapped.ILLUMINA.bwa.CHS.low_coverage.20120522.bam         2012-11-15 22:21  169M  
    1. HG00403.unmapped.ILLUMINA.bwa.CHS.low_coverage.20120522.bam.bai     2012-11-15 22:21   29K  
    1. HG00403.unmapped.ILLUMINA.bwa.CHS.low_coverage.20120522.bam.bas  
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00403/sequence_read/
    1. ERR013127.filt.fastq.gz               2010-06-26 01:35  8.8M  
    1. ERR013127_1.filt.fastq.gz             2010-06-26 01:35  1.7G  
    1. ERR013127_2.filt.fastq.gz             2010-06-26 01:35  1.6G  
    1. ERR015743.filt.fastq.gz               2010-07-26 07:27  102M  
    1. ERR015743_1.filt.fastq.gz             2010-07-26 07:27  1.6G  
    1. ERR015743_2.filt.fastq.gz             2010-07-26 07:27  1.5G  
    1. ERR016002.filt.fastq.gz               2010-08-09 08:04   22M  
    1. ERR016002_1.filt.fastq.gz             2010-08-09 08:04  1.1G  
    1. ERR016002_2.filt.fastq.gz             2010-08-09 08:04  1.3G 
    ..... about 54 lines
    1. SRR100444.filt.fastq.gz               2011-02-22 04:42  616K  
    1. SRR100444_1.filt.fastq.gz             2011-02-22 04:42  196M  
    1. SRR100444_2.filt.fastq.gz             2011-02-22 04:42  199M  
    1. SRR100445.filt.fastq.gz               2011-02-27 08:41  571K  
    1. SRR100445_1.filt.fastq.gz             2011-02-27 08:41  199M  
    1. SRR100445_2.filt.fastq.gz             2011-02-27 08:41  202M  
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/technical/reference/GRCh38_reference_genome/README.20150309.GRCh38_full_analysis_set_plus_decoy_hla
1. 

# HIGH COVERAGE

1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/

    1.                                             -   
    1. ERR3239276/                                2019-04-24 10:12    -   
    1. ERR3239277/                                2019-04-24 10:40    -   
    1. ERR3239278/                                2019-04-24 10:51    -   
    1. ERR3239279/                                2019-04-24 11:02    -   
    1. ERR3239280/                                2019-04-24 11:02    -   
    1. ERR3239281/                                2019-04-24 11:02    -   
    1. ERR3239282/                                2019-04-24 12:25    -   
    1. ERR3239283/                                2019-04-24 12:25    -   
    1. ERR3239284/                                2019-04-24 12:25    -   
    1. ERR3239285/                                2019-04-24 12:25    -   
    1. ERR3239286/                                2019-04-24 12:25    -   
    .....
    1. 
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/data/ERR3239276/
    1. NA06985.final.cram                              2019-04-08 10:42   16G  
    1. NA06985.final.cram.crai                         2019-04-24 11:02  1.4M  
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/1000G_2504_high_coverage.sequence.index
    1. #ENA_FILE_PATH	MD5SUM	RUN_ID	STUDY_ID	STUDY_NAME	CENTER_NAME	SUBMISSION_ID	SUBMISSION_DATE	SAMPLE_ID	SAMPLE_NAME	POPULATION	EXPERIMENT_ID	INSTRUMENT_PLATFORM	INSTRUMENT_MODEL	LIBRARY_NAME	RUN_NAME	INSERT_SIZE	LIBRARY_LAYOUT	PAIRED_FASTQ	READ_COUNT	BASE_COUNT	ANALYSIS_GROUP
    1. ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR323/ERR3239480/NA12718.final.cram	923ca8ff7d4cd65fddd28e855e5f173d	ERR3239480	ERP114329	30X whole genome sequencing coverage of the 2504 Phase 3 1000 Genome samples.	NYGC	ERA1783081	2019-03-22 00:00	SRS000632	NA12718	CEU	ERX3266855	ILLUMINA	Illumina NovaSeq 6000	NA12718	ena-RUN-NYGC-22-03-2019-12:30:23:667-205	450	PAIRED		753247448	112987117200	high_cov
    1. ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR324/ERR3241665/HG00403.final.cram	775634aa3e867f0ed255e2b0c64b5a84	ERR3241665	ERP114329	30X whole genome sequencing coverage of the 2504 Phase 3 1000 Genome samples.	NYGC	ERA1783865	2019-03-22 00:00	SRS008598	HG00403	CHS	ERX3268680	ILLUMINA	Illumina NovaSeq 6000	HG00403	ena-RUN-NYGC-22-03-2019-12:39:40:869-7	450	PAIRED		727457436	109118615400	high_cov
    1. ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR324/ERR3241666/HG00404.final.cram	a89652586e64fc4dddad9f487d04d038	ERR3241666	ERP114329	30X whole genome sequencing coverage of the 2504 Phase 3 1000 Genome samples.	NYGC	ERA1783865	2019-03-22 00:00	SRS008599	HG00404	CHS	ERX3268681	ILLUMINA	Illumina NovaSeq 6000	HG00404	ena-RUN-NYGC-22-03-2019-12:39:40:869-8	450	PAIRED		688651840	103297776000	high_cov
    1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/data/ERR3241665/
    1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/data/ERR3241665/
    1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/data/ERR3241665/
    1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/data/ERR3241665/
    1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/data/ERR3241665/
    1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/data/ERR3241665/
    1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/data/ERR3241665/
    1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/data/ERR3241665/
        1. HG00403.final.cram                              2019-04-18 23:36   14G  
        1. HG00403.final.cram.crai                         2019-04-24 10:59  1.3M 
        1. HG00403.final.cram                              2019-04-18 23:36   14G  
        1. HG00403.final.cram.crai                         2019-04-24 10:59  1.3M 
        1. HG00403.final.cram                              2019-04-18 23:36   14G  
        1. HG00403.final.cram.crai                         2019-04-24 10:59  1.3M 
        1. HG00403.final.cram                              2019-04-18 23:36   14G  
        1. HG00403.final.cram.crai                         2019-04-24 10:59  1.3M 
        1. HG00403.final.cram                              2019-04-18 23:36   14G  
        1. HG00403.final.cram.crai                         2019-04-24 10:59  1.3M 
        1. HG00403.final.cram                              2019-04-18 23:36   14G  
        1. HG00403.final.cram.crai                         2019-04-24 10:59  1.3M 
        1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00403/
        1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00403/
        1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00403/
        1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00403/
        1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00403/
        1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00403/


## History pdf presentation
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/technical/working/20120229_tutorial_docs/G1K_tutorial_brief_history_20120217.pdf


## README File Content

1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/technical/reference/GRCh38_reference_genome/README.20150309.GRCh38_full_analysis_set_plus_decoy_hla
1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/technical/reference/GRCh38_reference_genome/README.20150309.GRCh38_full_analysis_set_plus_decoy_hla

The reference fasta file in this directory was unpacked from Heng Li's bwakit-0.7.12:

http://sourceforge.net/projects/bio-bwa/files/bwakit/bwakit-0.7.12_x64-linux.tar.bz2/download

Heng's fasta file is based on GRCh38 full analysis set 

ftp://ftp.ncbi.nlm.nih.gov/genbank/genomes/Eukaryotes/vertebrates_mammals/Homo_sapiens/GRCh38/seqs_for_alignment_pipelines/GCA_000001405.15_GRCh38_full_plus_hs38d1_analysis_set.fna

The only addition is the HLA sequences.

This fasta file includes:

Chromosomes:
chr{chromosome number or name}
e.g. chr1 or chrX
chrM for the mitochondrial genome.

Unlocalized scaffolds:
chr{chromosome number or name}_{sequence_accession}v{sequence_version}_random
e.g. chr17_GL000205v2_random

Unplaced scaffolds:
chrUn_{sequence_accession}v{sequence_version}
e.g. chrUn_GL000220v1

Alternate loci scaffolds:
chr{chromosome number or name}_{sequence_accession}v{sequence_version}_alt
e.g. chr6_GL000250v2_alt

Patch scaffolds:
chr{chromosome number or name}_{sequence_accession}v{sequence_version}_patch
e.g. chrY_KN196487v1_patch 

Decoy sequences:
chrUn_{sequence_accession}v{sequence_version}_decoy
e.g. chrUn_KN707606v1_decoy

HLA sequences:
HLA-{id}
e.g. HLA-A*02:53N
 
This directory also contains some other files

20150713_location_of_centromeres_and_other_regions.txt contains a list of all the modelled centromeric sequence     
that GRCh8 contains as well as the location of the PAR on chrY and the hetrochromatic sequence on chr7.

This was sourced from ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_assembly_regions.txt on the 13th July

The other_mapping_resources contains a set of vcf files for snps and indels for base quality recalibration
and indel realignment for GRCh38 mapping pipelines.

There is more explanation about these files in

ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/other_mapping_resources/README.20150504.other_mapping_resources