{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1000 genomes ftp\n",
    "\n",
    "1. found in [https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/](https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/)\n",
    "1. Also https://www.internationalgenome.org/data/\n",
    "1. https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/\n",
    "1. Amazon S3 : s3://1000genomes\n",
    "1. [https://www.internationalgenome.org/](https://www.internationalgenome.org/)\n",
    "\n",
    "### from \n",
    "1.[https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/](https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/)\n",
    "1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00096/sequence_read/\n",
    "1. ```\n",
    "Parent Directory                                           -   \n",
    "SRR062634.filt.fastq.gz               2010-08-28 04:15   24M  \n",
    "SRR062634_1.filt.fastq.gz             2010-08-28 04:15  1.8G  \n",
    "SRR062634_2.filt.fastq.gz             2010-08-28 04:15  1.8G  \n",
    "SRR062635.filt.fastq.gz               2010-08-28 12:32  9.6M  \n",
    "SRR062635_1.filt.fastq.gz             2010-08-28 12:32  1.8G  \n",
    "SRR062635_2.filt.fastq.gz             2010-08-28 12:32  1.8G  \n",
    "SRR062641.filt.fastq.gz               2010-08-31 08:13  8.9M  \n",
    "SRR062641_1.filt.fastq.gz             2010-08-31 08:13  1.9G  \n",
    "SRR062641_2.filt.fastq.gz             2010-08-31 08:13  1.8G  \n",
    "SRR077487.filt.fastq.gz               2011-01-07 08:28   14M  \n",
    "SRR077487_1.filt.fastq.gz             2011-01-07 08:28  1.8G  \n",
    "SRR077487_2.filt.fastq.gz             2011-01-07 08:28  1.9G  \n",
    "SRR081241.filt.fastq.gz               2011-01-24 06:33   14M  \n",
    "SRR081241_1.filt.fastq.gz             2011-01-24 06:33  1.8G  \n",
    "SRR081241_2.filt.fastq.gz             2011-01-24 06:33  1.9G  \n",
    "```\n",
    "1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG00699/sequence_read/\n",
    "1. Parent Directory    \n",
    "```                                       -   \n",
    "ERR020268.filt.fastq.gz               2010-10-25 02:17   47M  \n",
    "ERR020268_1.filt.fastq.gz             2010-10-25 02:17  6.8G  \n",
    "ERR020268_2.filt.fastq.gz             2010-10-25 02:17  6.9G  \n",
    "ERR031925.filt.fastq.gz               2011-04-08 03:34   21M  \n",
    "ERR031925_1.filt.fastq.gz             2011-04-08 03:34  4.5G  \n",
    "ERR031925_2.filt.fastq.gz             2011-04-08 03:34  4.7G  \n",
    "```\n",
    "1. [https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG04209/sequence_read/]https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/phase3/data/HG04209/sequence_read/\n",
    "1. \n",
    "```\n",
    "Parent Directory                                           -   \n",
    "ERR250908.filt.fastq.gz               2013-04-07 10:22   39M  \n",
    "ERR250908_1.filt.fastq.gz             2013-04-07 10:22  2.3G  \n",
    "ERR250908_2.filt.fastq.gz             2013-04-07 10:22  2.4G  \n",
    "ERR250909.filt.fastq.gz               2013-04-05 15:15   41M  \n",
    "ERR250909_1.filt.fastq.gz             2013-04-05 15:15  2.3G  \n",
    "ERR250909_2.filt.fastq.gz             2013-04-05 15:15  2.3G  \n",
    "ERR251511.filt.fastq.gz               2013-04-09 09:07  9.9M  \n",
    "ERR251511_1.filt.fastq.gz             2013-04-09 09:07  5.9G  \n",
    "ERR251511_2.filt.fastq.gz             2013-04-09 09:07  6.1G  \n",
    "ERR251512.filt.fastq.gz               2013-04-09 08:58   14M  \n",
    "ERR251512_1.filt.fastq.gz             2013-04-09 08:58  5.9G  \n",
    "ERR251512_2.filt.fastq.gz             2013-04-09 08:58  6.0G  \n",
    "```\n",
    "\n",
    "1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/README.sequence_data\n",
    "\n",
    "1. https://ftp.sra.ebi.ac.uk/vol1/run/ERR398/ERR3988761/ HG00405 cram crai files ; 14 GB, 1.3 MB\n",
    "1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/additional_698_related/data/ERR3988761/\n",
    "\n",
    "\n",
    "## CRAM\n",
    "\n",
    "CRAM is a more highly compressed alternative to the BAM and SAM DNA sequence alignment file formats.\n",
    "1. [https://davetang.org/muse/2014/09/26/bam-to-cram/](https://davetang.org/muse/2014/09/26/bam-to-cram/)\n",
    "1. [38p14?](https://github.com/broadinstitute/gatk/blob/master/src/test/resources/large/Homo_sapiens_assembly38.fasta.gz)\n",
    "\n",
    "## Defintion CRAM/CRAI\n",
    "CRAM is a format designed by the European Bioinformatics Institute. It uses reference based compression, meaning that only base calls that differ to a designated reference sequence need to be stored. \n",
    "See http://www.ebi.ac.uk/ena/software/cram-toolkit for the EBIâ€™s documentation. Typically CRAM achieves 40-50% space saving over the alternative BAM format. The CRAM specification can be found at the http://samtools.github.io/hts-specs/ page along with related formats.\n",
    "\n",
    "## SAM format\n",
    "\n",
    "1. [https://ubpred.org/asm-file-format.html](https://ubpred.org/asm-file-format.html)\n",
    "\n",
    "\n",
    "# high coverage ref to bam to cram to bam.cram etc\n",
    "\n",
    "1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/1000G_2504_high_coverage/working/20201028_3202_raw_GT_with_annot/20201028_1000G_2020Oct26_NYGC_JointGenotyping_README.pdf\n",
    "    1. 20201028_1000G_2020Oct26_NYGC_JointGenotyping_README.pdf                                          2021-12-09 08:39   82K  \n",
    "    1. 20201028_CCDG_14151_B01_GRM_WGS_2020-08-05_chr1.recalibrated_variants.annotated.clinical.txt      2021-12-09 08:39  681K  \n",
    "    1. 20201028_CCDG_14151_B01_GRM_WGS_2020-08-05_chr1.recalibrated_variants.annotated.coding.txt        2021-12-09 08:39   35M  \n",
    "    1. 20201028_CCDG_14151_B01_GRM_WGS_2020-08-05_chr1.recalibrated_variants.annotated.coding_rare.txt   2021-12-09 08:40   34M  \n",
    "    1. 20201028_CCDG_14151_B01_GRM_WGS_2020-08-05_chr1.recalibrated_variants.annotated.txt               2021-12-09 08:53  4.5G  \n",
    "    1. 20201028_CCDG_14151_B01_GRM_WGS_2020-08-05_chr1.recalibrated_variants.annotated.vcf.gz            2021-12-09 09:05  1.7G  \n",
    "    1. 20201028_CCDG_14151_B01_GRM_WGS_2020-08-05_chr1.recalibrated_variants.annotated.vcf.gz.tbi        2021-12-09 09:05  219K  \n",
    "    1. 20201028_CCDG_14151_B01_GRM_WGS_2020-08-05_chr1.recalibrated_variants.vcf.gz                      2021-12-12 12:10  130G  \n",
    "    1. 20201028_CCDG_14151_B01_GRM_WGS_2020-08-05_chr1.recalibrated_variants.vcf.gz.tbi   \n",
    "\n",
    "\n",
    "## Refernece genome\n",
    "\n",
    "1. https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/technical/reference/GRCh38_reference_genome/\n",
    "1. https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/\n",
    "1. content folder\n",
    "    1. [TXT]\t20150713_location_of_centromeres_and_other_regions.txt\t2015-07-13 13:33\t1.3K\t \n",
    "    1. [   ]\tGRCh38_full_analysis_set_plus_decoy_hla-extra.fa\t2015-03-09 17:03\t7.9M\t \n",
    "    1. [   ]\tGRCh38_full_analysis_set_plus_decoy_hla.dict\t2015-03-27 19:49\t739K\t \n",
    "    1. [   ]\tGRCh38_full_analysis_set_plus_decoy_hla.fa\t2015-03-09 17:03\t3.0G\t \n",
    "    1. [   ]\tGRCh38_full_analysis_set_plus_decoy_hla.fa.alt\t2015-03-09 17:03\t476K\t \n",
    "    1. [   ]\tGRCh38_full_analysis_set_plus_decoy_hla.fa.amb\t2015-03-09 17:03\t20K\t \n",
    "    1. [   ]\tGRCh38_full_analysis_set_plus_decoy_hla.fa.ann\t2015-03-09 17:03\t438K\t \n",
    "    1. [   ]\tGRCh38_full_analysis_set_plus_decoy_hla.fa.bwt\t2015-03-09 17:03\t3.0G\t \n",
    "    1. [   ]\tGRCh38_full_analysis_set_plus_decoy_hla.fa.fai\t2015-03-09 17:03\t151K\t \n",
    "    1. [   ]\tGRCh38_full_analysis_set_plus_decoy_hla.fa.pac\t2015-03-09 17:03\t767M\t \n",
    "    1. [   ]\tGRCh38_full_analysis_set_plus_decoy_hla.fa.sa\t2015-03-09 17:03\t1.5G\t \n",
    "    1. [   ]\tREADME.20150309.GRCh38_full_analysis_set_plus_decoy_hla\t2015-07-13 13:33\t1.9K\t \n",
    "    1. [DIR]\tother_mapping_resources/\t2016-12-08 16:12\t-\t \n",
    "\n",
    "\n",
    "## [Genome Reference Consortium](https://www.ncbi.nlm.nih.gov/grc)\n",
    "### Monday, May 9, 2022\n",
    "### GRCh38.p14 is now released!\n",
    "\n",
    "1. [Genome 38.p14 is not released](https://genomeref.blogspot.com/2022/05/grch38p14-is-now-released.html)\n",
    "GRCh38.p14 (GCA_000001405.29/GCF_000001405.40), the latest update to the human reference assembly, has been released! It adds 69 new patch scaffolds, 51 of which are FIX patches that update sequences on the GRCh38 reference chromosomes or alternate loci, while 18 are NOVEL patches, providing new alternate representations for complex genomic regions that are inadequately represented by a single sequence. Two previously released FIX patches were also updated. With this release, the reference assembly contains a total of 250 patch scaffolds (164 FIX, 90 NOVEL).\n",
    "\n",
    "30 of the patches included in this release include genome updates made in support of the MANE project, a joint NCBI-EBI effort to produce a minimal set of matching RefSeq and Ensembl transcripts of protein coding genes, creating a matched pair of transcripts but retaining their respective identifiers. Read more about MANE effort in their recent Nature publication. The corresponding patch updates to the reference assembly involved changes addressing normal human variation as well as correcting errors in the underlying component sequences. \n",
    "\n",
    "Of the 53 FIX patches in GRCh38.p14, 23 of these correct errors in individual assembly component sequences, resulting in updates to 12 gene representations (Table 1). 20 are variation-related updates, 12 of which provide the coding allele for 13 polymorphic pseudogenes that are non-coding on the corresponding GRCh38 chromosomes (Table 2). Additionally, 2 provided sequence updates at chromosomal loci where it's unclear if the GRCh38 sequence is in error or a rare haplotype. Patch scaffolds in GRCh38.p14 close 6 gaps in the reference assembly, and extend sequence into one other gap. 4 of the closed gaps are located within chromosomes, while the remaining 2 patch scaffolds closed \"pre-telomeric\" gaps, extending the sequence of the chromosome into the telomeric repeats.\n",
    "\n",
    "\n",
    "## Annotation release\n",
    "\n",
    "[https://www.ncbi.nlm.nih.gov/refseq/annotation_euk/Homo_sapiens/GCF_000001405.40-RS_2023_10/](https://www.ncbi.nlm.nih.gov/refseq/annotation_euk/Homo_sapiens/GCF_000001405.40-RS_2023_10/)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
